File: 其他项目/nornir-vsr/frontend/src/api/nornir.ts
import client from "./client";

export type CommandType =
  | "display"
  | "config"
  | "connectivity"
  | "multiline"
  | "config_download";

export interface CommandPayload {
  hosts?: string[];
  commandType: CommandType;
  command?: string;
  commands?: string[];
  useTiming?: boolean;
}

export interface CommandResult {
  host: string;
  logId?: number | null;
  snapshotId?: number | null;
  commandType: CommandType;
  command: string;
  result: string;
  failed: boolean;
  exception?: string | null;
  executedAt: string;
  outputPath?: string | null;
}

export const executeCommand = async (payload: CommandPayload): Promise<CommandResult[]> => {
  const { data } = await client.post<CommandResult[]>("/nornir/commands", payload, {
    // 前端不设超时，完全由后端控制
    timeout: 0,
  });
  return data;
};

export interface CommandHistoryQuery {
  limit?: number;
  host?: string;
  commandType?: CommandType;
  includeConfigDownload?: boolean;
}

export const fetchCommandHistory = async (
  params: CommandHistoryQuery = {}
): Promise<CommandResult[]> => {
  const { data } = await client.get<CommandResult[]>("/nornir/commands/history", {
    params,
    timeout: 0,
  });
  return data;
};

export const deleteCommandHistory = async (logId: number): Promise<void> => {
  await client.delete(`/nornir/commands/history/${logId}`, { timeout: 0 });
};

export interface ConfigSnapshotDetail {
  id: number;
  host: string;
  site?: string | null;
  command: string;
  content: string;
  executedAt: string;
  filePath?: string | null;
}

export const fetchConfigSnapshot = async (snapshotId: number): Promise<ConfigSnapshotDetail> => {
  const { data } = await client.get<ConfigSnapshotDetail>(`/nornir/config-snapshots/${snapshotId}`, {
    timeout: 0,
  });
  return data;
};

export interface ConfigSnapshotSummary {
  id: number;
  host: string;
  site?: string | null;
  command: string;
  executedAt: string;
  filePath?: string | null;
}

export interface ConfigSnapshotQuery {
  limit?: number;
  host?: string;
  site?: string;
}

export const fetchConfigSnapshots = async (
  params: ConfigSnapshotQuery = {}
): Promise<ConfigSnapshotSummary[]> => {
  const { data } = await client.get<ConfigSnapshotSummary[]>("/nornir/config-snapshots", {
    params,
    timeout: 0,
  });
  return data;
};

export const fetchLatestConfigSnapshots = async (
  params: ConfigSnapshotQuery = {}
): Promise<ConfigSnapshotSummary[]> => {
  const { data } = await client.get<ConfigSnapshotSummary[]>("/nornir/config-snapshots/latest", {
    params,
    timeout: 0,
  });
  return data;
};

export const deleteConfigSnapshot = async (snapshotId: number): Promise<void> => {
  await client.delete(`/nornir/config-snapshots/${snapshotId}`, { timeout: 0 });
};

export const deleteConfigSnapshots = async (snapshotIds: number[]): Promise<number> => {
  const { data } = await client.delete(`/nornir/config-snapshots/batch`, {
    data: { ids: snapshotIds },
    timeout: 0,
  });
  return (data?.deleted as number) ?? 0;
};

export const downloadConfigSnapshot = async (snapshotId: number): Promise<Blob> => {
  const response = await client.get(`/nornir/config-snapshots/${snapshotId}/download`, {
    responseType: "blob",
    timeout: 0,
  });
  return response.data;
};

File: 其他项目/nornir-vsr/backend/schemas/nornir.py
"""Nornir 任务相关模型。"""
from datetime import datetime
from enum import Enum
from typing import List, Optional

from pydantic import BaseModel, ConfigDict, Field, model_validator


class CommandType(str, Enum):
    DISPLAY = "display"
    CONFIG = "config"
    CONNECTIVITY = "connectivity"
    MULTILINE = "multiline"
    CONFIG_DOWNLOAD = "config_download"


class NornirCommandRequest(BaseModel):
    hosts: Optional[List[str]] = None
    command_type: CommandType = Field(alias="commandType")
    command: Optional[str] = None
    commands: Optional[List[str]] = None
    use_timing: bool = Field(default=False, alias="useTiming")

    @model_validator(mode="after")
    def validate_payload(self) -> "NornirCommandRequest":
        if self.command_type in {CommandType.DISPLAY, CommandType.CONNECTIVITY}:
            if not self.command or not self.command.strip():
                raise ValueError("command is required for the selected command type")
            self.command = self.command.strip()

        if self.command_type in {CommandType.CONFIG, CommandType.MULTILINE}:
            commands: List[str] = []
            if self.commands:
                commands = [cmd.strip() for cmd in self.commands if cmd.strip()]
            elif self.command:
                commands = [line.strip() for line in self.command.splitlines() if line.strip()]

            if not commands:
                raise ValueError("commands are required for the selected command type")

            self.commands = commands

        if self.command_type == CommandType.CONFIG_DOWNLOAD:
            # 固定使用 display current-configuration
            self.command = "display current-configuration"

        return self


class NornirCommandResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    host: str
    log_id: Optional[int] = Field(default=None, alias="logId")
    snapshot_id: Optional[int] = Field(default=None, alias="snapshotId")
    command_type: CommandType = Field(alias="commandType")
    command: str
    result: str
    failed: bool
    exception: Optional[str] = None
    executed_at: datetime = Field(alias="executedAt")
    output_path: Optional[str] = Field(default=None, alias="outputPath")


class CommandLogEntry(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    host_name: str
    site: Optional[str] = None
    command: str
    command_type: CommandType
    result: Optional[str] = None
    success: bool
    exception: Optional[str] = None
    output_path: Optional[str] = None
    executed_at: datetime


class ConfigSnapshotSummary(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: int
    host: str
    site: Optional[str] = None
    command: str
    executed_at: datetime = Field(alias="executedAt")
    file_path: Optional[str] = Field(default=None, alias="filePath")


class ConfigSnapshotDetail(ConfigSnapshotSummary):
    content: str
